crates/node-bindings/src/parcel/parcel.rs

```diff
@@ -1,62 +1,138 @@
+use std::io;
 use std::path::PathBuf;
+use std::sync::mpsc::channel;
+use std::sync::mpsc::Sender;
 use std::sync::Arc;
+use std::thread;
 
 use napi::Env;
 use napi::JsObject;
 use napi_derive::napi;
-use parcel::file_system::FileSystemRef;
 use parcel::Parcel;
 use parcel::ParcelOptions;
 use parcel_plugin_rpc::nodejs::RpcHostNodejs;
 
 use crate::file_system::FileSystemNapi;
 
+enum ParcelMessage {
+  RpcPing {
+    response: Sender<()>,
+  },
+  FsReadToString {
+    path: PathBuf,
+    response: Sender<io::Result<String>>,
+  },
+  FsIsFile {
+    path: PathBuf,
+    response: Sender<bool>,
+  },
+  FsIsDir {
+    path: PathBuf,
+    response: Sender<bool>,
+  },
+}
+
+enum ParcelResponse {}
+
 #[napi]
 pub struct ParcelNapi {
-  internal: Arc<Parcel>,
+  tx_parcel: Sender<ParcelMessage>,
 }
 
 #[napi]
 impl ParcelNapi {
   #[napi(constructor)]
   pub fn new(env: Env, options: JsObject) -> napi::Result<Self> {
+    // Debugging Instrumentation
     let _ = tracing_subscriber::fmt::try_init();
-
     let thread_id = std::thread::current().id();
     tracing::trace!(?thread_id, "parcel-napi initialize");
 
-    let rpc_host_nodejs = RpcHostNodejs::new();
-
-    let mut fs = None::<FileSystemRef>;
+    // Wrap the JavaScript-supplied FileSystem
+    let fs = FileSystemNapi::from_options(&env, &options)?;
 
-    if options.has_named_property("fs")? {
-      let fs_raw: JsObject = options.get_named_property("fs")?;
-      fs.replace(Arc::new(FileSystemNapi::new(&env, fs_raw)?));
-    }
+    // Set up Nodejs plugin bindings
+    let rpc_host_nodejs = RpcHostNodejs::new(&env, options.get_named_property("rpc")?)?;
 
+    // Initialize Parcel
     let parcel = Parcel::new(ParcelOptions {
       fs,
       rpc: Some(Arc::new(rpc_host_nodejs)),
     });
 
-    Ok(Self {
-      internal: Arc::new(parcel),
-    })
+    // Run Parcel within its own thread
+    let (tx_parcel, rx_parcel) = channel::<ParcelMessage>();
+    thread::spawn(move || {
+      while let Ok(msg) = rx_parcel.recv() {
+        if match msg {
+          ParcelMessage::FsReadToString { path, response } => {
+            response.send(parcel.fs.read_to_string(&path)).is_err()
+          }
+          ParcelMessage::FsIsFile { path, response } => {
+            response.send(parcel.fs.is_file(&path)).is_err()
+          }
+          ParcelMessage::FsIsDir { path, response } => {
+            response.send(parcel.fs.is_dir(&path)).is_err()
+          }
+          ParcelMessage::RpcPing { response } => response
+            .send(parcel.rpc.as_ref().unwrap().ping().unwrap())
+            .is_err(),
+        } {
+          return;
+        }
+      }
+    });
```

Comment:
Running Parcel in its own thread is required because `RpcHostNodejs` will block the main thread waiting for a response. After a while this can be flattened again though any methods we run will need to be on their own thread - something like this:

```rust
#[napi]
impl ParcelNapi {
  #[napi]
  pub fn build(&self) {
    let tsfn = /**/
    let parcel = self.parcel.clone();
    thread::spawn(move || {
       tsfn.call(Ok(parcel.build()));
    });
  }
}
```

Reply:
Actually the `#[napi] async fn` runs on a Tokio worker thread so it should be fine